File: source/com/barrybecker4/common/math1/Range.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.common.math1;

/**
 * Manage a double precision range.
 *
 * @author Barry Becker
 */
public class Range {

    private double min;
    private double max;

    /** Default constructor */
    public Range() {
        min = Double.MAX_VALUE;
        max = -Double.MAX_VALUE;
    }

    /**
     * init with min and max values of the range.
     * @param minimum min value for range
     * @param maximum max value for range
     */
    public Range(double minimum, double maximum) {
        if (minimum > maximum) {
            throw new IllegalArgumentException("min greater than max");
        }
        min = minimum;
        max = maximum;
    }

    /** Copy constructor */
    public Range(Range range) {
        min = range.getMin();
        max = range.getMax();
    }

    /**
     * @return Returns the Max.
     */
    public double getMax() {
        return max;
    }

    /**
     * @return Returns the axisMin.
     */
    public double getMin() {
        return min;
    }

    /**
     *Extend this range by the range argument.
     */
    public void add(Range range) {
        add(range.getMin());
        add(range.getMax());
    }

    /**
     * Extend this range by the value argument.
     */
    public void add(double value) {
        if (value < min) {
            min = value;
        }
        if (value > max) {
            max = value;
        }
    }

    /**
     * The extend of the range.
     * @return the max minus the min.
     */
    public double getExtent() {
        if (min > max) {
            return Double.NaN;
        }
        return (max - min);
    }

    /**
     * @return true if the range is completely contained by us.
     */
    public boolean inRange(Range range) {
        return range.getMin() >= getMin() && range.getMax() <= getMax();
    }

    /**
     * @param value value to normalize on unit scale.
     * @return  normalized value assuming 0 for min. 1 for max.
     */
    public double mapToUnitScale(double value) {
        double range = getExtent();
        if (range == 0) {
            return 0;
        }
        return (value - getMin()) / getExtent();
    }

    public String toString() {
        return  this.getMin() + " to " + this.getMax(); //NON-NLS
    }

    @Override
    public boolean equals(Object obj) {
        if (!(obj instanceof Range))
            return false;
        Range otherRange = (Range) obj;
        return (this.getMin() == otherRange.getMin() && this.getMax() == otherRange.getMax());
    }

    public int hashCode() {
        Double sum = (this.getMin() + this.getMax())/2.0;
        return sum.hashCode();
    }
}

File: scala-source/com/barrybecker4/common/math/Range.scala
/* Copyright by Barry G. Becker, 2000-2018. Licensed under MIT License: http://www.opensource.org/licenses/MIT */
package com.barrybecker4.common.math

/**
  * A double precision range.
  * @author Barry Becker
  */
case class Range(min: Double = Double.NaN, max: Double = Double.NaN) {

  assert(min.isNaN && max.isNaN || min <= max)

  def this(range: Range) { this(range.min, range.max) }

  /** @return new range extending the current one by specified range. */
  def add(range: Range): Range = add(range.min).add(range.max)

  /** @return a new range extending this range by the value argument */
  def add(value: Double): Range = {
    if (min.isNaN) Range(value, value)
    else if (value < min) Range(value, max)
    else if (value > max) Range(min, value)
    else this
  }

  /** @return the max minus the min. */
  def getExtent: Double = {
    if (min.isInfinite || max.isInfinite || (min.isNaN && max.isNaN)) return Double.NaN
    max - min
  }

  /** @return true if the range is completely contained by us. */
  def inRange(range: Range): Boolean = range.min >= min && range.max <= max

  /** @param value value to normalize on unit scale.
    * @return normalized value assuming 0 for min. 1 for max.
    */
  def mapToUnitScale(value: Double): Double = {
    val range = getExtent
    if (range == 0) return 0
    (value - min) / getExtent
  }

  override def toString: String = this.min + " to " + this.max
}